### PR TITLE
[gitlab] pre-build pulumi go program

### PR DESCRIPTION
**Pull request by pducolin**
What does this PR do?
---------------------

* Build `main.go` pulumi program before running integration tests

Which scenarios this will impact?
-------------------

None

Motivation
----------

Running pulumi on the CI requires big runners, both in e2e tests in `datadog-agent` and here when calling invoke tasks to create remote instances, in `integration_tests`. 

@AliDatadog  few months ago noticed how the high mem consumption in e2e tests is requested at compile time. This happens both when running e2e tests in `datadog-agent` and when running integration_tests in test-infra-definitions. 

[I recently added](https://github.com/DataDog/test-infra-definitions/pull/916) `fakeintake` to `integration_tests` as a `pulumi-aws` bump broke its usage in `datadog-agent` e2e tests. This further increased the pipeline time to ~40 minutes, as tests ran sequentially. In an attempt to reduce the CI time here [I ran tests in parallel](https://github.com/DataDog/test-infra-definitions/pull/923). 

The tests run the `inv create-[vm|docker|kind]` tasks: when running concurrently they each build `main.go`, which requires a huge amount of RAM. I increased the `KUBERNETES_MEMORY_LIMIT` to the maximum allowed of `32GB`. 

This got OOM in a [recent PR](https://github.com/DataDog/test-infra-definitions/pull/915) where we add more dependencies.

@KSerrania , @KevinFairise2 and myself investigated what we could do to allow running tests without starving the CI. Building once in a `16GB` runner and then test in // in a `6GB` runner is my suggested solution. 

Additional Notes
----------------

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -50,14 +69,12 @@ integration-testing:
...
-    - go test ./integration-tests -v -timeout 0s
+    # execute test from dist directory to use the generated binary
+    - go test ./integration-tests -v -timeout 0s -workingDir=dist
   variables:
     E2E_PUBLIC_KEY_PATH: /tmp/agent-integration-test-ssh-key.pub
     E2E_PRIVATE_KEY_PATH: /tmp/agent-integration-test-ssh-key
     E2E_KEY_PAIR_NAME: e2e-integration-test-ssh-key
-    KUBERNETES_MEMORY_REQUEST: 24Gi
-    KUBERNETES_MEMORY_LIMIT: 30Gi
-    KUBERNETES_CPU_REQUEST: 12
```
Comment (KevinFairise2) [inline .gitlab-ci.yml:60]: **👏 praise:** ‏
That's very nice if thanks to your change the test execution fits on the default very small runner

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -30,11 +30,30 @@ build-runner-image:
...
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64:v38375519-53773733
+  tags: ["arch:amd64"]
+  rules:
+    - when: on_success
+  script:
+    - go build -o dist/main -gcflags="all=-c=6" main.go 
+  artifacts:
+    paths:
+      - dist/main
+    expire_in: "1 day"
```
Comment (KevinFairise2) [inline .gitlab-ci.yml:44]: Do you think we should add a warning if we try to rerun the integration test and the prebuilt artifact is no longer available?

**Review by KevinFairise2 — APPROVED** [review]
Very cool! 🚀

**Review by vboulineau — APPROVED** [review]

File: .gitlab-ci.yml (hunk 2 above)
Comment (pducolin) [reply .gitlab-ci.yml:44]: Yes, good idea. I can add a check at job time to ensure that the `dist/main` is still there, and recommend re-running the prev job on failure

Comment (pducolin) [reply .gitlab-ci.yml:44]: Added

**Pull request merged**